### PR TITLE
Friendship V2

### PR DESCRIPTION
**Pull request by balerum03**
The application now creates two row when the Friendship is confirmed (confirmed == true), this is done with the intention to have a mutual friendship register in the database.

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -5,5 +5,10 @@
     <ul class="posts">
       <%= render @posts %>
     </ul>
+    <%= @user_friends.each do |u| %>
+      <% if !find_friendship(current_user.id, u.id).nil? %>
+        <%= link_to 'hello', user_path(u.id) %>
+      <% end %>
+    <% end %>
```
Comment (teekaytech) [inline app/views/users/show.html.erb]: Kindly move logic away from views. View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb) 
Check [this](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) guide to get started with helpers.

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've made laudable progress in this milestone. Most features required seems to be working.__

__You've been informed that during the code review, you will get hints about implementing mutual friendships in the most efficient way, I'm here to do justice to that. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] When friendship is confirmed, you can create an additional row by adding this code snippet into your friendship model, like this:

```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [x] Thanks to the snippet above, we can find user friends only by `user_id` in the friendships table:

```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

- [x] Also, User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

- [x] At the same time, User#friend_requests can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

- [x] In order to get all posts written by a user and their friends, you can create and use a simple method in the User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

- [x] You did not have any integration tests in your project. Kindly add

**COMMENTS**
Please note that the explanations above concerning implementing mutual friendship are to enable you to see an efficient implementation of mutual friendship. 

In my opinion, it does not mean that you should completely refactor yours to use the method stated above. However, you are advised to see how you can use it to improve your own version of the implementation.


* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Comment by bmuhamood** [comment]
- All changes requested by TSE are added to project. 
- Fixed all linters errors.
- All tests added previously are simultaneously working with this project version.

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

## Required Changes ♻️ 

- [x] update the timeline_posts method from your posts_controller to use a method from the User model

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

_this is better than the current one as it uses less number of operations to achieve the same goal_ 

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Comment by balerum03** [comment]
The suggested code in the last review was pretty much already implemented. We tried to make changes to it but had no success.
The code was already working fine so we returned the code to the last working stage.

**Review by Dandush03 — APPROVED** [review]
# Status: Project Approved :rocket: 
## Great work so far, keep up the good job
# Happy Coding :nerd_face: :comet:

**Pull request merged**